utils/local_store.ml

```diff
@@ -0,0 +1,74 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                        Frederic Bour, Tarides                          *)
+(*                         Thomas Refis, Tarides                          *)
+(*                                                                        *)
+(*   Copyright 2020 Tarides                                               *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+type ref_and_reset =
+  | Table : { ref: 'a ref; init: unit -> 'a } -> ref_and_reset
+  | Immutable : { ref: 'a ref; mutable snapshot: 'a } -> ref_and_reset
+
+type bindings = {
+  mutable refs: ref_and_reset list;
+  mutable frozen : bool;
+  is_bound: bool ref
+}
+
+let global_bindings =
+  { refs = []; is_bound = ref false; frozen = false }
+
+let is_bound () = !(global_bindings.is_bound)
+
+let reset () =
+  assert (is_bound ());
+  List.iter (function
+    | Table { ref; init } -> ref := init ()
+    | Immutable { ref; snapshot } -> ref := snapshot
+  ) global_bindings.refs
+
+let s_table create size =
+  let init () = create size in
+  let ref = ref (init ()) in
+  assert (not global_bindings.frozen);
+  global_bindings.refs <- (Table { ref; init }) :: global_bindings.refs;
+  ref
+
+let s_ref k =
+  let ref = ref k in
+  assert (not global_bindings.frozen);
+  global_bindings.refs <-
+    (Immutable { ref; snapshot = k }) :: global_bindings.refs;
+  ref
+
+type slot = Slot : { ref : 'a ref; mutable value : 'a } -> slot
+type scope = { slots: slot list; scope_bound : bool ref }
+
+let fresh () =
+  let slots =
+    List.map (function
+      | Table { ref; init } -> Slot {ref; value = init ()}
+      | Immutable r ->
+          if not global_bindings.frozen then r.snapshot <- !(r.ref);
+          Slot { ref = r.ref; value = r.snapshot }
+    ) global_bindings.refs
+  in
+  global_bindings.frozen <- true;
+  { slots; scope_bound = global_bindings.is_bound }
```

Comment:
I don't understand the implementation of `fresh` and `with_scope`.

What I would expect is that `fresh` copies the (initial) values of `global_bindings` in an independent, fresh new store, that is not affected by further mutations to `global_bindings`, and then that `with_scope` would someone replace the current value of `global_bindings` by the ones in the store snapshot argument. This is not at all what the implementation does.

In the implementation, the mutable state is *shared* between `global_bindings` and the "fresh" store -- along with a copy in the `value` fields of the field state we would get after calling `reset` on the global store. `with_scope` then sets this shared mutable state to the content of the `value` field, which is thus related to calling `reset` but a bit different.

For example, if I call `fresh ()` twice in a row, then I get two different `scope`s whose `scope_bound` field is a shared boolean reference: at any point in time they are either both "bound" or none of them is "bound". Naively it looks like this could be implemented with a global boolean reference instead. More generally, what is the point of being able to handle several different "scopes", if in fact they all share the same mutable state? I'm missing something.

Reply:
> For example, if I call `fresh ()` twice in a row, then I get two different `scope`s whose `scope_bound` field is a shared boolean reference: at any point in time they are either both "bound" or none of them is "bound". Naively it looks like this could be implemented with a global boolean reference instead.

You are correct, with the new (simplified) API a single global boolean will be enough. I'll simplify that.

> More generally, what is the point of being able to handle several different "scopes", if in fact they all share the same mutable state? I'm missing something.

But they do not (apart from the "bound" field, as you rightly pointed out).
A scope is basically a list of pairs, where one element of the pair is a global reference, and the other is the value this reference should be set to when the scope becomes active.
Indeed every scope will talk about the same global references, but they will each have a different value that the reference should be set to.

> along with a copy in the `value` fields of the field state we would get after calling `reset` on the global store. `with_scope` then sets this shared mutable state to the content of the `value` field, which is thus related to calling `reset` but a bit different.

That is only part of the story. What you describe is true for the first call to with `with_scope`, it is equivalent to calling reset. But at the end of that call, the value of the global references is then snapshoted once again: but only for that scope (whose `value` field gets updated). A later call to `with_scope` will set the global references to these values, not the initial ones.